content/blog/2025-07-14-user-defined-parquet-indexes.md

```diff
@@ -0,0 +1,545 @@
+
+The example requires **arrow‑rs v55.2.0** or later, which includes the new “buffered write” API ([apache/arrow-rs#7714]) that keeps the internal byte count in sync so you can append index bytes immediately after data pages.
+
+[apache/arrow-rs#7714]: https://github.com/apache/arrow-rs/pull/7714  
+
+This example is intentionally simple for clarity, but you can adapt the same approach for any index type or data types. The high-level design is:
+
+1. **Choose or define your index payload** (e.g., bitmap, Bloom filter, sketch, distinct values list, etc.).
+
+2. **Serialize your index to bytes** and append them into the Parquet file body before writing the footer.
+
+3. **Record the index location** by adding a key/value entry (e.g., `"my_index_offset" -> "<byte‑offset>"`) in the Parquet footer metadata.
+
+4. **Extend DataFusion** with a custom `TableProvider` (or wrap the existing Parquet provider).
+
+
+The `TableProvider` simply reads the footer metadata to discover the index offset, seeks to that offset and deserializes the index, and then uses the index to speed up processing (e.g., skip files, row groups, data pages, etc.).
+
+The resulting Parquet files remain fully compatible with other tools such as DuckDB and Spark, which simply ignore the unknown index bytes and key/value metadata.
+
+
+### Introduction to Distinct Value Indexes
+
+---
+
+A **distinct value index** stores the unique values of a specific column. This type of index is effective for columns with a small number of distinct values and can be used to quickly skip files that do not match the query. These indexes are popular in several engines, such as the ["set" Skip Index in ClickHouse] and the [Distinct Value Cache] in InfluxDB 3.0.
+
+["set" Skip Index in ClickHouse]: https://clickhouse.com/docs/optimize/skipping-indexes#set
+[Distinct Value Cache]: https://docs.influxdata.com/influxdb3/enterprise/admin/distinct-value-cache/
+
+For example, if the files contain a column named `Category` like this:
+
+<table style="border-collapse:collapse;">
+  <tr>
+    <td style="border:1px solid #888;padding:2px 6px;"><b><code>Category</code></b></td>
+  </tr>
+  <tr>
+    <td style="border:1px solid #888;padding:2px 6px;"><code>foo</code></td>
+  </tr>
+  <tr>
+    <td style="border:1px solid #888;padding:2px 6px;"><code>bar</code></td>
+  </tr>
+  <tr>
+    <td style="border:1px solid #888;padding:2px 6px;"><code>...</code></td>
+  </tr>
+  <tr>
+    <td style="border:1px solid #888;padding:2px 6px;"><code>baz</code></td>
+  </tr>
+  <tr>
+    <td style="border:1px solid #888;padding:2px 6px;"><code>foo</code></td>
+  </tr>
+</table>
+
+The distinct value index will contain the values `foo`, `bar`, and `baz`. Using traditional min/max statistics would store the minimum (`bar`) and maximum (`foo`) values, which would not allow quickly skipping this file for a query like `SELECT * FROM t WHERE Category = 'bas'` as `bas` is between `bar` and `foo`.
```

Comment:
So the DV index(is it correct to call it index btw?) gives an opportunity to identify values within a parquet file or specific row group without sorting? To make min/max work effectively the sorting is needed

Reply:
Added some more explain in 
https://github.com/apache/datafusion-site/pull/79/commits/134207f06df5b84e2c845a7afa20433f552bd6de

Thank you @comphead !